src/mm-broadband-modem-mbim.c

```diff
@@ -710,16 +710,50 @@ pin_query_unlock_retries_ready (MbimDevice *device,
             NULL,
             &remaining_attempts,
             &error)) {
+        MMModemLock lock;
         MMUnlockRetries *retries;
 
+        lock = mm_modem_lock_from_mbim_pin_type (pin_type);
         retries = mm_unlock_retries_new ();
+
+        /* If PIN1 is disabled and we have tried to enable it with a wrong PIN,
+         * the modem would have indicated the number of remaining attempts for
+         * PIN1 (unless PUK1 is engaged) in the response to the failed
+         * MBIM_CID_PIN set operation. Thus, MMSimMbim would have updated
+         * MMIfaceModem's MMUnlockRetries with information about PIN1.
+         *
+         * However, a MBIM_CID_PIN query may be issued (e.g. MMBaseSim calls
+         * mm_iface_modem_update_lock_info()) after the MBIM_CID_PIN set
+         * operation to query the number of remaining attempts for a PIN type.
+         * Unfortunately, we can't specify a particular PIN type in a
+         * MBIM_CID_PIN query. The modem may not reply with information about
+         * PIN1 if PIN1 is disabled. When that happens, we would like to
+         * preserve our knowledge about the number of remaining attempts for
+         * PIN1. Here we thus carry over any existing information on PIN1 from
+         * MMIfaceModem's MMUnlockRetries if the MBIM_CID_PIN query reports
+         * something other than PIN1. */
+        if (lock != MM_MODEM_LOCK_SIM_PIN) {
+            MMIfaceModem *self;
+            MMUnlockRetries *previous_retries;
+            guint previous_sim_pin_retries;
+
+            self = g_task_get_source_object (task);
+            previous_retries = mm_iface_modem_get_unlock_retries (self);
```

Comment:
@cbchan this seems kinda icky; instead, could we just always keep an internal cache of SIM_PIN retry count in 'priv' instead?  Then do the same thing as here, call mm_unlock_retries_set() with that data.

Reply:
@dcbw @aleksander0m Currently, MMIfaceModem obtains MMUnlockRetries through MmGdbusModem. We can certainly keep a cache of MMUnlockRetries in MMBroadbandModemMbimPrivate. My only concern is that it seems a bit confusing to follow the code when both MmGdbusModem and MMBroadbandModemMbimPrivate hold a copy of MMUnlockRetries, which may or may not be in sync (though that may not be an issue). WDYT?

Reply:
We really do what @dcbw suggests for all the other items everywhere, i.e. keep the values internally and propagate them to the iface, only in that direction. So it probably is worth doing that as well here.

Reply:
That sounds good to me. Will update the patch set.

Reply:
Updated patch set. PTAL